File: LastSemesterInfo/AYDA Laboratorio/AYDA_Examen_2_Grafos_(A2026)/student_version/windows/python/template.py
import sys

from graph import UndirectedGraph, DirectedGraph, Node
from graph_algorithms import GraphTraversals, GraphProperties, GraphTopological

# ========================================================================
# HELPER FUNCTION (Provided for you)
# ========================================================================
def find_node(graph, val: str) -> Node:
    """Finds a node in the graph by its string value."""
    for node in graph.nodes:
        if node.value == val:
            return node
    return None

# ========================================================================
# EXERCISE 1: The Main Event (Wrestling Rivalries)
# ========================================================================
def classify_wrestlers(roster: UndirectedGraph) -> tuple[bool, set, set]:
    """
    TODO: Implement your logic here.
    Return a tuple: (is_possible: bool, faces: set, heels: set)
    """
    faces = set()
    heels = set()

    return False, faces, heels

# ========================================================================
# EXERCISE 2: The Phantom Heist (Strict Waypoints)
# ========================================================================
def plan_heist(mansion: UndirectedGraph, entry: Node, vault: Node, escape: Node) -> list[Node]:
    """
    TODO: Implement your logic here.
    Return a list of nodes representing the shortest path: Entry -> Vault -> Escape.
    If impossible, return an empty list [].
    """
    return []

# ========================================================================
# EXERCISE 3: Network Triage (Prioritization)
# ========================================================================
def get_repair_priority(network: UndirectedGraph) -> list[list[Node]]:
    """
    TODO: Implement your logic here.
    Return a list of lists (isolated network segments), sorted by size (descending).
    """
    return []

# ========================================================================
# EXERCISE 4: Parallel Build Orchestrator
# ========================================================================
def get_parallel_build_schedule(project_modules: DirectedGraph) -> list[list[Node]]:
    """
    TODO: Implement your logic here.
    Return a list of lists (stages). Each stage must be sorted alphabetically by node.value.
    If the graph has a cycle, return an empty list.
    """
    return []

# ========================================================================
# I/O BOILERPLATE (DO NOT MODIFY BELOW THIS LINE)
# This section communicates with the Automated Grader.
# ========================================================================
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Error: No exercise specified.")
        sys.exit(1)

    exercise = sys.argv[1]

    with open("solution.txt", "w") as sol:
        if exercise == "ex1":
            roster = UndirectedGraph.load_from_file("test_graph.txt", str, float)
            possible, faces, heels = classify_wrestlers(roster)

            sol.write("1\n" if possible else "0\n")
            if possible:
                sol.write(" ".join([n.value for n in faces]) + "\n")
                sol.write(" ".join([n.value for n in heels]) + "\n")

        elif exercise == "ex2":
            mansion = UndirectedGraph.load_from_file("test_graph.txt", str, float)

            with open("params.txt", "r") as params:
                lines = params.read().splitlines()
                e_val, v_val, esc_val = lines[0], lines[1], lines[2]

            entry = find_node(mansion, e_val)
            vault = find_node(mansion, v_val)
            escape = find_node(mansion, esc_val)

            route = plan_heist(mansion, entry, vault, escape)
            sol.write(" ".join([n.value for n in route]) + "\n")

        elif exercise == "ex3":
            network = UndirectedGraph.load_from_file("test_graph.txt", str, float)
            segments = get_repair_priority(network)

            for seg in segments:
                sol.write(" ".join([n.value for n in seg]) + "\n")

        elif exercise == "ex4":
            dag = DirectedGraph.load_from_file("test_graph.txt", str, float)
            stages = get_parallel_build_schedule(dag)

            if not stages:
                sol.write("CYCLE\n")
            else:
                for stage in stages:
                    sol.write(" ".join([n.value for n in stage]) + "\n")

File: LastSemesterInfo/AYDA Laboratorio/AYDA_Examen_2_Grafos_(A2026)/student_version/windows/rust/template.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead, Write};
use std::collections::HashSet;
use std::rc::Rc;

use crate::graph::{DirectedGraph, UndirectedGraph, Graph, Node};
use crate::graph_algorithms::{GraphTraversals, GraphProperties, GraphTopological};

// ========================================================================
// HELPER FUNCTION (Provided for you)
// Finds a node in the graph by its string value.
// ========================================================================
fn find_node(graph: &impl Graph<String, f64>, val: &str) -> Option<Rc<Node<String>>> {
    graph.get_nodes().iter().find(|n| n.value == val).cloned()
}

// ========================================================================
// EXERCISE 1: The Main Event (Wrestling Rivalries)
// ========================================================================
pub fn classify_wrestlers(
    roster: &UndirectedGraph<String, f64>,
    faces: &mut HashSet<Rc<Node<String>>>,
    heels: &mut HashSet<Rc<Node<String>>>
) -> bool {
    // TODO: Implement your logic here using BFS or DFS.
    // Return true if possible, and populate 'faces' and 'heels'.

    false
}

// ========================================================================
// EXERCISE 2: The Phantom Heist (Strict Waypoints)
// ========================================================================
pub fn plan_heist(
    mansion: &UndirectedGraph<String, f64>,
    entry: Rc<Node<String>>,
    vault: Rc<Node<String>>,
    escape: Rc<Node<String>>
) -> Vec<Rc<Node<String>>> {
    // TODO: Implement your logic here.
    // Return the shortest path. If impossible, return an empty Vec.

    Vec::new()
}

// ========================================================================
// EXERCISE 3: Network Triage (Prioritization)
// ========================================================================
pub fn get_repair_priority(
    network: &UndirectedGraph<String, f64>
) -> Vec<Vec<Rc<Node<String>>>> {
    // TODO: Implement your logic here.
    // Return a list of isolated network segments, sorted by size (descending).

    Vec::new()
}

// ========================================================================
// EXERCISE 4: Parallel Build Orchestrator
// ========================================================================
pub fn get_parallel_build_schedule(
    project_modules: &DirectedGraph<String, f64>
) -> Vec<Vec<Rc<Node<String>>>> {
    // TODO: Implement your logic here.
    // Tip: Use `GraphProperties::has_cycle` first to avoid panicking!
    // Return stages sorted alphabetically by node.value. If a cycle exists, return empty Vec.

    Vec::new()
}

// ========================================================================
// I/O BOILERPLATE (DO NOT MODIFY BELOW THIS LINE)
// This section communicates with the Automated Grader.
// ========================================================================
fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Error: No exercise specified.");
        std::process::exit(1);
    }

    let exercise = &args[1];
    let mut sol = File::create("solution.txt")?;

    match exercise.as_str() {
        "ex1" => {
            let roster = UndirectedGraph::<String, f64>::load_from_file("test_graph.txt")?;
            let mut faces = HashSet::new();
            let mut heels = HashSet::new();

            let possible = classify_wrestlers(&roster, &mut faces, &mut heels);

            if possible {
                writeln!(sol, "1")?;
                for n in faces { write!(sol, "{} ", n.value)?; }
                writeln!(sol)?;
                for n in heels { write!(sol, "{} ", n.value)?; }
                writeln!(sol)?;
            } else {
                writeln!(sol, "0")?;
            }
        },
        "ex2" => {
            let mansion = UndirectedGraph::<String, f64>::load_from_file("test_graph.txt")?;

            let params_file = File::open("params.txt")?;
            let mut lines = io::BufReader::new(params_file).lines();

            let e_val = lines.next().unwrap().unwrap();
            let v_val = lines.next().unwrap().unwrap();
            let esc_val = lines.next().unwrap().unwrap();

            let entry = find_node(&mansion, &e_val).unwrap();
            let vault = find_node(&mansion, &v_val).unwrap();
            let escape = find_node(&mansion, &esc_val).unwrap();

            let route = plan_heist(&mansion, entry, vault, escape);
            for n in route { write!(sol, "{} ", n.value)?; }
            writeln!(sol)?;
        },
        "ex3" => {
            let network = UndirectedGraph::<String, f64>::load_from_file("test_graph.txt")?;
            let segments = get_repair_priority(&network);

            for seg in segments {
                for n in seg { write!(sol, "{} ", n.value)?; }
                writeln!(sol)?;
            }
        },
        "ex4" => {
            let dag = DirectedGraph::<String, f64>::load_from_file("test_graph.txt")?;
            let stages = get_parallel_build_schedule(&dag);

            if stages.is_empty() {
                writeln!(sol, "CYCLE")?;
            } else {
                for stage in stages {
                    for n in stage { write!(sol, "{} ", n.value)?; }
                    writeln!(sol)?;
                }
            }
        },
        _ => eprintln!("Unknown exercise: {}", exercise),
    }

    Ok(())
}